src/OrleansProviders/Streams/Memory/MemoryAdapterFactory.cs

```diff
@@ -0,0 +1,127 @@
+            this.adapterConfig = new MemoryAdapterConfig(providerName);
+            grainFactory = (GrainFactory)serviceProvider.GetService(typeof(GrainFactory));
+            adapterConfig.PopulateFromProviderConfig(providerConfig);
+            this.streamQueueMapper = new HashRingBasedStreamQueueMapper(adapterConfig.TotalQueueCount, adapterConfig.StreamProviderName);
+
+            // 10 meg buffer pool.  10 1 meg blocks
+            this.bufferPool = new FixedSizeObjectPool<FixedSizeBuffer>(adapterConfig.CacheSizeMb, () => new FixedSizeBuffer(1 << 20));
+        }
+
+        public Task<IQueueAdapter> CreateAdapter()
+        {
+            return Task.FromResult<IQueueAdapter>(this);
+        }
+
+        public IQueueAdapterCache GetQueueAdapterCache()
+        {
+            return this;
+        } 
+
+        public IStreamQueueMapper GetStreamQueueMapper()
+        {
+            return streamQueueMapper;
+        }
+
+        public IQueueAdapterReceiver CreateReceiver(QueueId queueId)
+        {
+            IQueueAdapterReceiver receiver = (IQueueAdapterReceiver)new MemoryAdapterReceiver(GetQueueGrain(queueId));
+            return receiver;
+        }
+
+        public async Task QueueMessageBatchAsync<T>(Guid streamGuid, string streamNamespace, IEnumerable<T> events, StreamSequenceToken token, Dictionary<string, object> requestContext)
+        {
+            QueueId queueId;
+            MemoryEventData eventData;
+            try
+            {
+                queueId = streamQueueMapper.GetQueueForStream(streamGuid, streamNamespace);
+                eventData = MemoryBatchContainer.ToMemoryEventData(streamGuid, streamNamespace, events, requestContext);
+                IMemoryStreamQueueGrain queueGrain = GetQueueGrain(queueId);
+                await queueGrain.Enqueue(eventData);
+            }
+            catch (Exception exc)
+            {
+                logger.Error((int)ProviderErrorCode.MemoryStreamProviderBase_QueueMessageBatchAsync, "Exception thrown in MemoryAdapterFactory.QueueMessageBatchAsync.", exc);
+                throw;
+            }
+        }
+
+        public IQueueCache CreateQueueCache(QueueId queueId)
+        {
+            return new MemoryPooledCache(bufferPool, logger.GetSubLogger("-Cache"));
+        }
+
+        public Task<IStreamFailureHandler> GetDeliveryFailureHandler(QueueId queueId)
+        {
+            return Task.FromResult(streamFailureHandler ?? (streamFailureHandler = new NoOpStreamDeliveryFailureHandler()));
+        }
+
+        private Guid GenerateDeterministicGuid(QueueId queueId)
+        {
+            // provider name hash code
+            JenkinsHash jenkinsHash = JenkinsHash.Factory.GetHashGenerator();
+            int providerNameGuidHash = (int)jenkinsHash.ComputeHash(providerName);
+
+            // get queueId hash code
+            uint queueIdHash = queueId.GetUniformHashCode();
```

Comment:
in order to trust this uniform hash code, please remove the overloads for `QueueId.GetQueueId` that do not take an explicit hash. The currently have no usages, so it's an easy removal, but will prevent naive usages in the future.